### PR TITLE
Webcrypto refactor, fix type of publicExponent

### PR DESCRIPTION
**Pull request by fhanau**
This addresses #48 and refactors the `EdDsaKey` class. 
- In terms of refactoring, we can combine the EdDsaKey and EdDsaCurveKey classes as they are largely identical, the subtle difference in the naming of the algorithm field can be handled in getAlgorithm() alone.
- For the issue I primarily changed the clone() method of RsaKeyAlgorithm to return the correct type for publicExponent instead of making more invasive changes, this should minimize the risk of introducing bugs. The change is behind the new `cryptoPreservePublicExponent` compat flag, the other changes are for adding a test and introducing additional scaffolding
- Some additional questions regarding potential refactoring came up while reviewing the code to look for any other situations where we might be returning the wrong type, see comments.

### CODE REVIEW

File: src/workerd/api/crypto.h [hunk 1]
```diff
@@ -229,7 +250,7 @@ class CryptoKey: public jsg::Object {
       KeyAlgorithm, AesKeyAlgorithm, HmacKeyAlgorithm, RsaKeyAlgorithm,
       EllipticKeyAlgorithm, ArbitraryKeyAlgorithm>;
```
Comment (fhanau) [inline src/workerd/api/crypto.h:251]: ArbitraryKeyAlgorithm isn't really used at the moment – do we still need to have it exported/included in AlgorithmVariant? Do we expect it to have future uses?

Comment (fhanau) [reply src/workerd/api/crypto.h:251]: GitHub doesn't let me select the right line but the hash field in ArbitraryKeyAlgorithm should be `jsg::Optional<kj::OneOf<KeyAlgorithm, kj::StringPtr>` right?

Comment (fhanau) [reply src/workerd/api/crypto.h:251]: Similarly, uint32_t or uint64_t should work better for saltLength in `AsymmetricKeyDetails`, there's no need for salt sizes beyond uint32_t.

Comment (jasnell) [reply src/workerd/api/crypto.h:251]: `uint32_t` is probably good.

**Review by jasnell — APPROVED** [review]

**Pull request merged**